### ISSUE TITLE
fl_call_main must translate command line arguments from wide char to utf8

### ISSUE THREAD

**Issue by MatthiasWM** [fltk/fltk#840]
see: https://groups.google.com/g/fltkcoredev/c/NPnnDvVIUCE/m/7sIRaHfrCAAJ

**Describe the bug**
It seems that fl_call_main does not translate command line parameters from Chinese wide characters to utf8 correctly.

**FLTK Version**
 - Version: probably 1.4 master

**Additional context**
Gonzalo Garramuño posted the following code as a suggestion to fix the issue.

```
#ifdef _WIN32

#    include <iostream>
#    include <locale>
#    include <codecvt>

// Function to convert wstring to UTF-8 string
std::string wstring_to_utf8(const std::wstring& ws)
{
std::wstring_convert<std::codecvt_utf8<wchar_t>> converter;
    return converter.to_bytes(ws);
}

int WinMain(
    HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLine, int
nCmdShow)
{
    // Convert the command line arguments to UTF-8
    int argc;
    char** argv;

    // Get the command line as a wide string
    LPWSTR* wideArgv = CommandLineToArgvW(GetCommandLineW(), &argc);

    // Convert each wide string argument to UTF-8
    argv = new char*[argc];
    for (int i = 0; i < argc; ++i)
    {
        argv[i] = strdup(wstring_to_utf8(wideArgv[i]).c_str());
    }

    // Free the wide string array
    LocalFree(wideArgv);

    int ret = main(argc, argv);

    // Cleanup allocated memory for argv
    for (int i = 0; i < argc; ++i)
    {
        free(argv[i]);
    }
    delete[] argv;

    return ret;
}

#endif
```

**Comment by Albrecht-S** [comment]
@MatthiasWM I didn't see yet that you created this issue and assigned it to yourself. I also started investigation and have now a working prototype: I could verify that the existing code **_does NOT work well_** and I could also verify that **_my new code works_**. Note that this code is not yet ready for production.

How much have you been working already, and should we maybe synchronize somehow to avoid duplicated work?

**Comment by Albrecht-S** [comment]
Note: I'm currently using `CommandLineToArgvW(GetCommandLineW(), &argc);` as proposed but I'm using `WideCharToMultiByte()` to convert the MultiByte string to UTF-8 which we are - surprisingly [1] - using elsewhere in our codebase.

[1] `WideCharToMultiByte()` is used in some isolated places, presumably because this was user-contributed code whereas we're using our own convertion functions elsewhere. I was about to look into this as well to consolidate our conversions from Window wide char to UTF-8 and vice versa.

**Comment by Albrecht-S** [comment]
FWIW, here is my current, already working code (this is incomplete, some variable declarations are missing here):
```C++
  // Convert the command line arguments to UTF-8
  int argc;
  char** argv;

  // Get the command line as a wide string
  LPWSTR* wideArgv = CommandLineToArgvW(GetCommandLineW(), &argc);

  /* Convert each wide string argument to UTF-8 */
  argv = malloc(argc * sizeof(void *));
  for (int i = 0; i < argc; ++i) {
    char strbuf[255];
    int ret = WideCharToMultiByte(CP_UTF8,              /* CodePage         */
                                  0,                    /* dwFlags          */
                                  wideArgv[i],          /* lpWideCharStr    */
                                  -1,                   /* cchWideChar      */
                                  strbuf,               /* lpMultiByteStr   */
                                  sizeof(strbuf),       /* cbMultiByte      */
                                  NULL,                 /* lpDefaultChar    */
                                  NULL);                /* lpUsedDefaultCha */
      argv[i] = fl_strdup(strbuf);
  }

  // Free the wide string array
  LocalFree(wideArgv);

  rc = main(argc, argv);

  // Cleanup allocated memory for argv
  for (int i = 0; i < argc; ++i) {
    free(argv[i]);
  }
  free((void *)argv);
```

I wonder if we should use `WideCharToMultiByte()` everywhere to convert Windows wide chars to UTF-8 (and it's opposite function for the other direction).

**Comment by MatthiasWM** [comment]
Ah, since there was no reply, I decided to make this Issue. Thanks for already looking at the code. It sure makes sense to use `WideCharToMultiByte()` if it is available on all versions of MSWindows. It would replace a whole page of handmade code.

**Comment by Albrecht-S** [comment]
@MatthiasWM OK, then I can take this issue and continue and we don't waste our time?

Meanwhile I could see that my mods work when built with Visual Studio but there seems to be no difference when built with MSYS2/MinGW-w64. I'll have another try with old MinGW. This needs more investigation. Sigh.

**Comment by MatthiasWM** [comment]
Yes, please continue. Thank you.

**Comment by Albrecht-S** [comment]
For the record:
- we're using `WideCharToMultiByte()` and `MultiByteToWideChar()` already in `src/Fl_Native_File_Chooser_WIN32.cxx` and `src/drivers/WinAPI/Fl_WinAPI_System_Driver.cxx`
- the documentation says: `Minimum supported client: Windows 2000 Professional [desktop apps | UWP apps]`.

I believe this is OK for our supported platforms, and nobody complained so far.

I decided to use `WideCharToMultiByte()` in `fl_call_main.c` directly. I'll check other usage of wide character conversion functions on Windows (Wide Char to UTF-8 and vice versa) later.

**Comment by Albrecht-S** [comment]
@ggarra13 Can you confirm that commit 7e8994c4a295e8709d4940656248c231de62a8a6 fixes the issue for you?

**Comment by ggarra13** [comment]
This week we are in the middle of a critical Presidential election (which will probably devaluate our currency) here in Argentina, so it will take me a while to look into it.

**Comment by Albrecht-S** [comment]
OK, no problem. I'll wait until you have time (the commit is online anyway).

**Comment by Albrecht-S** [comment]
**_Clarification:_** the previous commit 7e8994c4a295e8709d4940656248c231de62a8a6 modifies the existing code that provides the `WinMain()` entry point for Windows programs and calls the more standard `main()` entry point as documented. This code does now convert the `argv[]` array transparently and reliably to UTF-8 before the `main()` program is activated. However, this code is only active for _Windows programs built using Visual Studio_.

Other tools like `MinGW` and `MSYS2`/`MinGW-w64` don't use Microsoft's special `WinMain()` entry and are not affected by this commit. This means that programs built with these build systems lack the new commandline argument conversion and accessing commandline arguments through the usual `argc`/`argv` mechanism is still not as expected. I'm investigating a potential solution.

**Comment by Albrecht-S** [comment]
I wrote:
> Other tools like MinGW and MSYS2/MinGW-w64 don't use Microsoft's special `WinMain()` entry and are not affected by this commit.

Commit 727bd94560ca9056cdbe40fe2e7923ed008b6eac addresses this issue by adding a new function `Fl::args_to_utf8()` that can be called on all platforms before argc and argv are used by the program. This function is a no-op on all platforms except windows (returns quickly) so you can call it at the beginning of `main()` on all platforms.

For details see docs at https://www.fltk.org/doc-1.4/classFl.html#af4ada403c5aaf71864cb159bdb5bdf62

**Comment by Albrecht-S** [comment]
@ggarra13 I would like to close this issue soon. Are there any news (test results) from you to be expected? I suggest to add a call to `Fl::args_to_utf8()` in your `main()` program to be safe on all platforms - or take a look at the documentation for alternatives.

**Comment by ggarra13** [comment]
Sorry.  I responded to your mail saying it was working (before the Fl::args_to_utf8() call), but somehow my answer emails don't make it to the github thread.  I'll try the new Fl::args_to_utf8() call today.

**Comment by ggarra13** [comment]
@Albrecht-S Okay, I tried it and found out I cannot use it as my application is a windows application and requires WinMain and cannot just depend on main.  Calling Fl::args_to_utf8() would do the conversion twice (unneeded) and the Fl::args_to_utf8 on the WinAPI driver seems to leak some minor memory, which my/and yours original solution did not.  So, I will stick to your original code and not use Fl::args_to_utf8().

**Comment by Albrecht-S** [comment]
@ggarra13 Thanks for testing.

You wrote: "I cannot use it as my application is a windows application and requires WinMain". As I found out "Windows application" is not exactly the right term in this context. In my experience you need WinMain() if you compile with VisualStudio (obviously because its runtime requires it), but WinMain() is not called if you build your **Windows** application e.g. with MinGW and/or MSYS/MinGW-w64 - although these build tools also create "Windows" and not console applications.

>  Calling Fl::args_to_utf8() would do the conversion twice (unneeded)

That is correct, I'm still looking for a better solution.

> Fl::args_to_utf8 on the WinAPI driver seems to leak some minor memory

This is also correct but can be fixed easily. However, it's not really a leak because the memory will be returned to the system at program exit. The implementation in `WinMain()` releases the memory after return from `main()` which doesn't make much difference.

> I will stick to your original code and not use Fl::args_to_utf8().

You're free to do that, thanks for testing nevertheless. But note that FLTK's own `WinMain()` implementation will only be used if you build your application with Visual Studio. If that's all you need and want to do then you're fine with not using `Fl::args_to_utf8()`.

Closing now.

**Comment by MatthiasWM** [comment]
Looking for a better solution

**Comment by MatthiasWM** [comment]
Ok, so I just pushed (new git client on Windows! ;-) ) a change that makes the buffer size for the string conversion dynamic. It's tested with VisualStudio and works with exotic UTF-8 characters from the catalog of alchemy. Freeing the buffers on exit is a waste of time IMHO. Is there anything else to do here? Maybe add the call to all our test apps?

I was thinking about `Fl::args(int argc, char **argv, int& i, Fl_Args_Handler cb = 0);`, and in a way, This would be a nice place to automatically do this call? Maybe in `Fl_Window::show(args, args);` if `Fl::args()` was not called?

If neither of these are called, FLTK can not apply its custom command line options, which is sad, because users may want to change the scheme or the colors. Worse yet, `Fl_Window::show(args, args);` sets the app colors, so if it is *not* called, the app will look different (a bit darker on default macOS, but changes may be very different on machines with custom color schemes). So it is generally a good idea to call `Fl_Window::show(args, args);`. 

Which brings me to another issue: if a user calls `Fl::args(args, args)`, he later still must call `Fl_Window::show(args, argv)` for some of the changes to be applied. This is done so that `Fl::args()` does not open the display. Nevertheless, the API is repetitive and IMHO unexpected.

Suggestion 1: we set a flag that indicates if `Fl_Window::show(args, argv)` was already called. If not, a first call to `Fl_Window::show()` (without argument list) will automatically be forwarded to `Fl_Window::show(args, argv)` with NULL arguments. That way, at least the system colors are applied. And if `Fl::args(args, args)`, those settings are applied.

Suggestion 2: we highly recommend that users call `Fl::args(args, args)`, or `Fl::args(args, args, int& i, Fl_Args_Handler cb)` so that end users can change schemes and colors, or `Fl::args(args, args, NO_FLTK_ARGS)` if the developer wants to suppress FLTK argument interpretation for some reason.

Suggestion 3: we remove the `Fl::args_to_utf8()` and make it part of `Fl::args(...)`.

**Comment by Albrecht-S** [comment]
> I just pushed ... a change that makes the buffer size for the string conversion dynamic.

:+1: Looks good, thanks.

But there is another instance of "the same code" in `src/fl_call_main.c` (which must be pure C code). Since you changed the one in the driver code, that one should IMHO be changed as well.

> Freeing the buffers on exit is a waste of time IMHO.

Yup, we shouldn't do this, it would be a waste of developers' time.
However, the converted strings have been free'd in `src/fl_call_main.c` almost since forever and I left this as-is. We should probably keep this.

> Is there anything else to do here?

No, I don't think so (except the mentioned changes in `src/fl_call_main.c`).

> Maybe add the call to all our test apps?

I would not do it for _all test apps_ but it could be useful for apps that really parse arguments (besides the FLTK standard args). For instance (IIRC) `test/pixmap_browser.cxx` that can take a PATH argument, and of course FLUID (if that's not already the case).

Everything else you wrote in your comment is something we could consider, but I wouldn't touch anything else before 1.4.0 (even not during 1.4.x unless we **_NEED_** to fix a bug). We should reserve further changes to 1.5.0 which can probably be released at some time in 2024. I'm optimistic.

**Comment by MatthiasWM** [comment]
Fixed in fl_call_main as well.

**Comment by Albrecht-S** [comment]
> Fixed in fl_call_main as well.

Sorry for nitpicking: there must not be C++ (`//`)  comments in C files. Although I admit that there was at least one in the version before your update (left by me). It's only necessary to change this for either old compilers or special compiler warnings like `-pedantic` or `-W` but it's in our CMP as well.

Fixed in commit e951db4c49a13106fe5628fde7bb031a13e82915

**Comment by Albrecht-S** [comment]
I confirm that the issue when building with MinGW has been fixed and I tried several of the modified test/demo programs as well by using (on Linux with cross-compiled FLTK) e.g. "wine bin/test/demo.exe -ti "äöü € µ 側経意責家方家閉":

![image](https://github.com/fltk/fltk/assets/26235500/3ba5e52b-2d61-42f2-9de0-e12f34286852)

I think we should close the issue now.
